### PR TITLE
LEDs: GPS fix, Sat count, RSSI & Flash

### PR DESCRIPTION
**Pull request by gaelj**
New LED functions: G & S
Isssue #709
Isssue #1384

### CODE REVIEW

**Comment by hydra** [comment]
There are compilation errors.  Please see build log details and add another commit that fixes the problems.

**Comment by gaelj** [comment]
Oops. It only appeared as a warning in my eclipse. Fixed.

**Comment by skaman82** [comment]
Great! Could you also add a Beacon mode? The configured LED should flash in a preselected color. Usefull as position lights.

**Comment by skaman82** [comment]
Here is a example: https://m.youtube.com/watch?v=neq82MD5v3A

**Comment by martinbudden** [comment]
@gaelj , test code is not passing. Can you fix that?

**Comment by gaelj** [comment]
@skaman82 I will be looking into the beacon mode now.
@martinbudden, @hydra, your feedback on the code is very welcome.

**Comment by martinbudden** [comment]
@gaelj , some feedback on your code, mostly to do with style.

1. I much prefer to declare variables at the point of usage, in my view it makes the code much easier to read. It also gives the opportunity to declare the variable const (if it is const), which aids understanding.
2. This is especially true of `for` loops, so, in your code, `for (int ledIndex = 0; ledIndex < ledCount;ledIndex++)...` rather than `for (ledIndex = 0; ledIndex < ledCount; ledIndex++)`
3. Unless there is good reason, I prefer to use `i` (and `j` and `k`) as loop indices, so in your case
` `for (int i = 0; i < ledCount; ++i)...` (I prefer the pre increment `++` operator - it's a good habit to use it since it is the more efficient form when it is an overloaded operator in C++)
4. I like the ternary conditional both for compactness and clarity, e.g. `gpsColor = STATE(GPS_FIX) ?&hsv_green : &hsv_red;`
5. In my view `continue` is "considered harmful", and should generally be avoided.
6. Double negatives should also generally be avoided (e.g. `if(!x) statement1 else statement2;` should be replaced by `if(x) statement2 else statement1;`. So in your code:
```
if (!(ledConfig->flags & LED_FUNCTION_GPS)) {
     continue;
}
setLedHsv(ledIndex, gpsColor);
```
should be replaced by
```
if ((ledConfig->flags & LED_FUNCTION_GPS)) {
    setLedHsv(ledIndex, gpsColor);
}
```
and similarly in the `applyLedRssiLayer` function.

**Comment by gaelj** [comment]
Fixed on my functions, however this coding style was used throughout the whole file.

**Comment by skaman82** [comment]
@gaelj Thanks, are you adding the new functions to the LED Tab interface on the configurator branch? It would be also great if https://github.com/cleanflight/cleanflight/pull/2053 would be configurable also in the GUI

**Comment by gaelj** [comment]
Of course @skaman82, this will be the next step once the cleanflight code is reviewed.

**Comment by ledvinap** [comment]
@martinbudden : 
ad 5/6: there is useful idiom for iterating with continue:
```
for(int i = 0; i < count; i++) {
   if(!filter1(item[i]))
     continue;
   if(!filter2(item[i]))
     continue;
   doSomething(item[i]);
   doSomethingElse();
}
```
This avoids deep nesting and is quite easy to read

**Comment by ledvinap** [comment]
@gaelj : On ARM processors (and probably most 32 bit processors + all desktops) it is more efficient to use `int` variables instead of shorted alternatives. 
- Local variables end up in registers and GCC emits lots of instructions to convert byte/halfword to word. 
- When passing arguments, values must be passed as 4 bytes, forcing sign or zero extension. This incurs overhead (and previous point applies inside function)
- When stored in memory, reading the value will take identical time (unless value is unaligned). Only advantage is saving RAM memory.

**Comment by gaelj** [comment]
see CF Configurator PR https://github.com/cleanflight/cleanflight-configurator/pull/379

**Comment by skaman82** [comment]
@gaelj just tested the blink feature. It works fantastic, thank you!

Also testing the GPS led function now. The strange thing is, that the LED is off with no fix (not red like you've stated in the docs). If I apply a color setting additionally it starts to blink in that selected color.
It starts working correctly if the first 3D fix is acquired. The, if the fix is lost, the led starts blinking red. 
It would be nice if the LED starts blinking red before the first fix is there.


Can not test the RSSI. Can someone else test that?

**Comment by gaelj** [comment]
I'll be looking into this new travis check failure + initialization issue you describe.
I don't have the means to test RSSI either but it's copy and paste from the throttle behaviour.
Would be nice to have some feedback though.

**Comment by gaelj** [comment]
@ledvinap your suggestion leads to an 18 byte flash overflow. Shall I revert or would you suggest something else ?

**Comment by skaman82** [comment]
Regarding travis check: it looks like the new code is to large for cc3d target.

Von meinem iPhone gesendet

**Comment by skaman82** [comment]
In the worst case you can leave the rssi out to save code space. I personally can't imagine how a rssi indication on a aircraf that is far away could be usefull at all.

**Comment by martinbudden** [comment]
@gaelj , I wouldn't worry about the 18 bytes overflow. I've got some PRs lined up that will save more than that amount of memory, so when they are merged this will build without errors.

**Comment by gaelj** [comment]
Yes I saw those @martinbudden. Excellent work. I did revert in the end just for the sake of not having the error and restore consistency with the coding style that is used throughout the file anyway.
Regarding RSSI not being very usefull, it was a suggestion brought up by @hydra so I would like to have his seconds thoughts on this one before.

**Comment by skaman82** [comment]
Testin now. The GPS mode LEDs are now red and blinking yellow before the first fix is acquired. If the fix is lost, the gps LEDs turn red for a moment and then they sometimes go just go completely OFF (happens not every time). They start blinking green again if the fix is back.

Suggestions:
• Better use orange color instead of yellow (it is hard to make out a difference between yellow and green in the sunlight)
• The GPS leds should be red again when fix is lost instead of going completely dark

**Comment by hydra** [comment]
please let me know when testing is complete so we can merge it.

thanks for code review guys.  the gps and blink features will be very useful indeed.

**Comment by gaelj** [comment]
The LEDs will always go black for a while if you wait enough, then it starts working again. I'm suspecting an overflow of the time variables but coding µs in a signed int32 still gives 35 minutes before overflow, and the blackness starts much before. So I'm still looking into it...

I've implemented the orange color. Expected behaviour is : 
- solid red when no satellites
- blinking orange when satellites but no fix
- blinking green when fix

**Comment by skaman82** [comment]
@gaelj ok, give me a quick note here when you have figure it out. I will test it again then.

**Comment by gaelj** [comment]
OK i've found the bug.... just another case of me being blind

**Comment by skaman82** [comment]
Just tested you changes before your last commit. In searching Mode The GPS LED is now blinking red:green (not orange). And if i pull the GPS the lights still go off

**Comment by skaman82** [comment]
Cortection. It goes of when i reattach The wires

**Comment by gaelj** [comment]
Please try thos last 2 commits. Seems pretty fine now to me (finally !)

**Comment by skaman82** [comment]
Still goes off when i pul and reattach The GPS wires. Strange thing is. First in searching Mode it blinks red/yellow. If it geht's a fix ist blinks YElLOW then i cut the gps connection and reattach it. The LEDs go off completely and starting blinking GREEN.

**Comment by skaman82** [comment]
@martinbudden could you please help debugging and test this if you have time? I can't test the RSSI function.

**Comment by martinbudden** [comment]
@skaman82  , unfortunately I don't have the facility to test the RSSI function.

**Comment by skaman82** [comment]
@martinbudden thanks for the quick feedback. Could you then take a look at the GPS LED function. The seems a bug somewhere causing strange behavior described above

**Comment by gaelj** [comment]
Over here when I disconnect the GPS unit, the red blink green a few more seconds then go solid red.
Then possibly blinking orange for one cycle and blinking green again. I don't get any dark pauses any more.

**Comment by skaman82** [comment]
I sadly do. @digitalentity may have some ideas. I am talking with him about this right now

**Comment by skaman82** [comment]
@gaelj please see improved GPS code here: https://gist.github.com/digitalentity/f4f021876653acb1f7f1525024388348

I tested this and this now works like it should. Pleas also add 
#include "sensors/sensors.h" at the top of the ledstrip.c file for this

tank you @digitalentity !

**Comment by digitalentity** [comment]
No problem, glad I could help :D

File: src/main/io/ledstrip.c [hunk 1]
```diff
@@ -924,15 +939,19 @@ void updateLedStrip(void)
 
     uint32_t now = micros();
 
-    bool indicatorFlashNow = (int32_t)(now - nextIndicatorFlashAt) >= 0L;
-    bool warningFlashNow = (int32_t)(now - nextWarningFlashAt) >= 0L;
-    bool rotationUpdateNow = (int32_t)(now - nextRotationUpdateAt) >= 0L;
+    bool indicatorFlashNow = (now >= nextIndicatorFlashAt);
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: `(int32_t)(now - nextIndicatorFlashAt) >= 0L` will work always (value is correct when micros wrap),  `(now >= nextIndicatorFlashAt);` will not work when now is  between `(~0 - nextIndicatorFlashAt) .. ~0`

Comment (ledvinap) [inline src/main/io/ledstrip.c]: BTW: there is 'cmp16' / 'cmp32' function to do the compare (`cmp32(now, nextIndicatorFlashAt) >= 0)`)

File: src/main/io/ledstrip.c [hunk 2]
```diff
@@ -1048,6 +1083,50 @@ bool parseColor(uint8_t index, const char *colorConfig)
     return ok;
 }
 
+/*
+ * Redefine a color in a mode.
+ * */
+bool setModeColor(uint8_t modeIndex, uint8_t modeColorIndex, uint8_t colorIndex)
+{
+    bool ok = true;
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: Why not just `return true;` / `return false;` ? `bool ok` is usually used to handle errors (setting it to false when something fails)., this is very confusing

File: src/main/io/ledstrip.c [hunk 3]
```diff
@@ -1048,6 +1083,50 @@ bool parseColor(uint8_t index, const char *colorConfig)
...
+    switch (modeIndex) {
+        case MODE_ORIENTATION:  modeColor = modeColors(MODE_ORIENTATION); break;
+        case MODE_HEADFREE:     modeColor = modeColors(MODE_HEADFREE); break;
+        case MODE_HORIZON:      modeColor = modeColors(MODE_HORIZON); break;
+        case MODE_ANGLE:        modeColor = modeColors(MODE_ANGLE); break;
+#ifdef MAG
+        case MODE_MAG:          modeColor = modeColors(MODE_MAG); break;
+#endif
+        case MODE_BARO:         modeColor = modeColors(MODE_BARO); break;
+        case SPECIAL:
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: Maybe use separate function to set special mode colors?

File: src/main/io/ledstrip.h [hunk 4]
```diff
@@ -68,9 +74,44 @@ typedef enum {
...
+} modeIds;
+
+typedef enum {
+    SC_FUNCTION_DISMARED = 0,
+    SC_FUNCTION_ARMED,
+    SC_FUNCTION_ANIMATION,
+    SC_FUNCTION_BACKGROUND
+} specialColorFunctionIds;
+
+typedef struct modeColorIndexes_s {
```
Comment (ledvinap) [inline src/main/io/ledstrip.h]: array + enum will work better

File: src/main/io/ledstrip.h [hunk 5]
```diff
@@ -68,9 +74,44 @@ typedef enum {
...
+    uint8_t north;
+    uint8_t east;
+    uint8_t south;
+    uint8_t west;
+    uint8_t up;
+    uint8_t down;
+} modeColorIndexes_t;
+
+typedef struct specialColorIndexes_s {
```
Comment (ledvinap) [inline src/main/io/ledstrip.h]: array + enum ?

File: src/main/io/ledstrip.h [hunk 6]
```diff
@@ -68,9 +74,44 @@ typedef enum {
...
 
+typedef enum {
+    MODE_ORIENTATION = 0,
+    MODE_HEADFREE,
+    MODE_HORIZON,
+    MODE_ANGLE,
+    MODE_MAG,
+    MODE_BARO,
+    SPECIAL
+} modeIds;
```
Comment (ledvinap) [inline src/main/io/ledstrip.h]: `modeId_e`

File: src/main/io/ledstrip.h [hunk 7]
```diff
@@ -68,9 +74,44 @@ typedef enum {
...
+
+typedef enum {
+    SC_FUNCTION_DISMARED = 0,
+    SC_FUNCTION_ARMED,
+    SC_FUNCTION_ANIMATION,
+    SC_FUNCTION_BACKGROUND
+} specialColorFunctionIds;
```
Comment (ledvinap) [inline src/main/io/ledstrip.h]: `specialColorFunctionId_e`

File: docs/LedStrip.md [hunk 8]
```diff
@@ -314,6 +333,47 @@ color 14 0,0,0
...
+- west
+- up
+- down
+
+Group 6 is used for special colors according to these ModelColorIndexes:
+
+- disarmed
+- armed
+- animation
+- bakground
```
Comment (ledvinap) [inline docs/LedStrip.md]: background

File: src/main/io/ledstrip.c [hunk 9]
```diff
@@ -56,6 +58,8 @@
 
 PG_REGISTER_ARR_WITH_RESET_FN(ledConfig_t, MAX_LED_STRIP_LENGTH, ledConfigs, PG_LED_STRIP_CONFIG, 0);
 PG_REGISTER_ARR_WITH_RESET_FN(hsvColor_t, CONFIGURABLE_COLOR_COUNT, colors, PG_COLOR_CONFIG, 0);
+PG_REGISTER_ARR_WITH_RESET_FN(modeColorIndexes_t, MODE_COUNT, modeColors, PG_MODE_COLOR_CONFIG, 0);
+PG_REGISTER_ARR_WITH_RESET_FN(specialColorIndexes_t, 1, specialColors, PG_SPECIAL_COLOR_CONFIG, 0);
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: Why array with 1 element?

File: src/main/io/ledstrip.c [hunk 10]
```diff
@@ -713,6 +663,44 @@ void applyLedWarningLayer(uint8_t updateNow)
...
+    for (uint8_t i = 0; i < ledCount; ++i) {
+
+        ledConfig = ledConfigs(i);
+
+        if (ledConfig->flags & LED_FUNCTION_GPS) {
+            setLedHsv(i, gpsColor);
+        }
+    }
+
+    if (updateNow) {
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]:
```
if (updateNow) {
  // handle delay first
  if(gpsPauseCounter) {
    gpsPauseCounter--;
  } else if(gpsFlashCounter >= GPS_numSat * 2) {
    gpsFlashCounter = 0;
    gpsPauseCounter = blinkPauseLength; // or use constant here
  }
}
```
and invert pause test / set initial pause if desired

File: src/main/io/ledstrip.c [hunk 11]
```diff
@@ -762,24 +750,21 @@ void applyLedIndicatorLayer(uint8_t indicatorFlashState)
...
+    for (uint8_t i = 0; i < ledCount; ++i) {
+        ledConfig = ledConfigs(i);
+        if (ledConfig->flags & flag) {
+            getLedHsv(i, &color);
 
-        int scaled = scaleRange(rcData[THROTTLE], PWM_RANGE_MIN, PWM_RANGE_MAX, -60, +60);
-        scaled += HSV_HUE_MAX;
-        color.h = (color.h + scaled) % HSV_HUE_MAX;
-        setLedHsv(ledIndex, &color);
+            int scaled = scaleRange(value, minRange, maxRange, -60, +60);
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: scaled may be moved outside of for loop

File: src/main/io/ledstrip.c [hunk 12]
```diff
@@ -945,14 +964,22 @@ void updateLedStrip(void)
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: quite a lot of counters and high code repetition. This may be handled as loop with template

File: src/main/io/serial_cli.c [hunk 13]
```diff
@@ -1253,6 +1255,109 @@ static void cliColor(char *cmdline)
...
+#ifdef MAG
+#else
+                    continue;
+#endif
+            }
+
+            for (j = 0; j < DIRECTIONS_COUNT; j++) {
+
+                switch (j) {
+                    case 0: colorIndex = modeColors(i)->north; break;
```
Comment (ledvinap) [inline src/main/io/serial_cli.c]: enum + array will avoid switch

File: src/main/io/serial_msp.c [hunk 14]
```diff
@@ -933,7 +933,7 @@ static bool processOutCommand(uint8_t cmdMSP)
         serialize8(rxConfig()->rssi_channel);
         serialize8(0);
 
-        serialize16(compassConfig()->mag_declination / 10);
```
Comment (ledvinap) [inline src/main/io/serial_msp.c]: this change is unrelated to LED, should be probably in separate PR

**Comment by gaelj** [comment]
Thanks guys. Seems the sensor check was not an issue on my side.
The general code cleanup was quite needed too IMO, although it's not always directly related to the PR.
I merged with the very latest commit of CF/master and solved all conflicts.
Now that the PR for cli command mode_color is merged, I will make the GPS-status colors configurable.

**Comment by skaman82** [comment]
@gaelj I think you should check this further. If I have 3 sats the blinking stops the leds a red for a while and the they go dark completely, then it blinks a few times yellow and the get just solid red without any blinking. Sometimes it even don't blink when the fix is there for maybee 15-20 seconds. It also goes completely dark if I pull and reattach the gps wires.

The code-snippet I posted above has worked very well tonight with no issues at all. Please have a closer look at it.

**Comment by gaelj** [comment]
It was working, but not up-to-date with the latest PR's.
I had to fix quite a lot of conflicts and then merging this PR gave some more conflicts.
Could you please give me your feedback on this commit ?

File: src/main/io/ledstrip.c [hunk 15]
```diff
@@ -660,32 +650,62 @@ void applyLedWarningLayer(uint8_t updateNow)
...
+    for (int i = 0; i < ledCount; ++i) {
+        const ledConfig_t *ledConfig = ledConfigs(i);
+
+        if (ledConfig->flags & LED_FUNCTION_GPS) {
+            setLedHsv(i, gpsColor);
+        }
+    }
+
+    if (updateNow) {
+        gpsFlashCounter++;
```
Comment (ledvinap) [inline src/main/io/ledstrip.c]: `gpsFlashCunter` must be incremented only when `gpsPauseCounter` is zero (sorry, I overlooked that)
```
        if (gpsPauseCounter) {
            gpsPauseCounter--;
        } else {
           gpsFlashCounter ++;
           if (gpsFlashCounter >= GPS_numSat * 2) {
            gpsFlashCounter = 0;
            gpsPauseCounter = blinkPauseLength;
        }
  }

**Comment by skaman82** [comment]
still no win. Green flashing just goes completely dark even when I have a fix (9 sats). If I replace the whole void applyLedGpsLayer(uint8_t updateNow) area with digitalentitys code I just works solid and good. I would suggest to just copy and paste, optimize later if necessary

**Comment by gaelj** [comment]
OK I will check. The commit will not solve it as I did not see your comment.

File: src/main/io/ledstrip.c (hunk 15 above)
Comment (gaelj) [inline src/main/io/ledstrip.c]: Saw that too ;-)

**Comment by gaelj** [comment]
@ledvinap could you please confirm if you are having problems as well ? Quite honestly it's working on my side and appart from the 2 latests commit (incrementation bugfix & configurable colors) the file contents of your version and mine match.
@skaman82 could you check you have the correct code and / or be more specific about the difference between both versions of the function ?

**Comment by gaelj** [comment]
OK, I got the line 19...

**Comment by ledvinap** [comment]
@gaelj : Sorry, I don't have gps or ledstrip at hand ... but the code should be simple enough ..

**Comment by gaelj** [comment]
OK, @skaman82 ...

**Comment by ledvinap** [comment]
@gaelj gpsFlashCounter is still increased during pause, so it will skip 2 sats

**Comment by skaman82** [comment]
Tested the state before your last commit. This is looking good, but after i while the green flashing just stops and the leds go dark. I just made a video while testing at the stage the green blinking stopped: https://www.youtube.com/watch?v=m1n5rbIq4Z8

**Comment by skaman82** [comment]
Tested f5e7bef but there is no difference. I'll try another FC tomorrow, maybe it is my SP racing board... At the mean time maybe @digitalentity could take a look and check what can cause the dropouts like in my video above?

**Comment by gaelj** [comment]
This should at least help.

**Comment by skaman82** [comment]
@gaelj Thank you, tested today with a new board and had no issues with the GPS LEDs - no matter how hart I tried to throw them off :) 

@hydra You wanted a heads-up when this is ready, I think it is now.

**Comment by martinbudden** [comment]
@skaman82 , the travis build is failing - that needs to be fixed before it can be included.

**Comment by skaman82** [comment]
@gaelj would you please check the travis errors?

**Comment by gaelj** [comment]
I'm glad it's finally working !!!
Obvisouly I will fix the Travis build - later today.
I hope this PR can still make it to 1.13 ?

**Comment by hydra** [comment]
yes if you fix this today it will be included, if not it'll be for 1.14

**Comment by gaelj** [comment]
@skaman82, did you test the working release on your SP F3 ? How does it behave ?

**Comment by skaman82** [comment]
Can test after work

**Comment by gaelj** [comment]
OK, everything's now fine except for CC3D build that overflowed by 62 bytes.
So I tried changing quite a few declarations from int to uint8_t, but now it's overflowing by 126 bytes !
@martinbudden, @ledvinap, I would very much welcome some hints...

**Comment by martinbudden** [comment]
@gaelj , as a general principle changing `int`s to `uint8_t`s is not going to save you ROM, as you've found out. The most efficient code is generated when variables are the "natural" size for the processor (that is, the same size as registers). Certainly for local variables `int` is the most efficient choice. For `static` and global variables it's more complicated though.

62 bytes isn't much, but then it's a lot when you absolutely need them. I'm afraid I don't have time to look at this more today, but a general piece of advice is that the easiest way to save ROM is to cut out functionality. I don't know your code that well, but if you could say cut out the RSSI functionality that might buy you the 62 bytes. You could put the RSSI functionality back in for 1.14, when you have more time to optimise the code.

**Comment by skaman82** [comment]
+1 Cut the RSSI

**Comment by gaelj** [comment]
RSSI is now just a function call to applyLedHue(), just the same as for throttle but with different parameters. So I doubt removing this function will change much.

**Comment by martinbudden** [comment]
@gaelj , yes, I wasn't hopeful.

**Comment by gaelj** [comment]
It's getting better though ! Maybe a deeper removal of rssi will do the trick.

**Comment by ledvinap** [comment]
@gaelj : you can probably save considerable ROM by optimizing tthis switch statement:
https://github.com/cleanflight/cleanflight/pull/2101/files#diff-bb48e8235ba420743f280df4cfa26af7R1063

**Comment by hydra** [comment]
@gaelj Is this ready now?

**Comment by skaman82** [comment]
testing now

**Comment by skaman82** [comment]
ready in 5 minutes

**Comment by gaelj** [comment]
I've plainly disabled GPS LEDs for CC3D. If this is OK for you it's good for me.

**Comment by gaelj** [comment]
@skaman82 is desperately waiting for his GPS fix :D

**Comment by skaman82** [comment]
:D works like a charm on my SPracing @hydra good to go!

**Comment by skaman82** [comment]
Big thanks to @gaelj @digitalentity @ledvinap @martinbudden

**Comment by gaelj** [comment]
And big thanks to @skaman82 ! I'm so glad this could make it to 1.13.

**Comment by martinbudden** [comment]
Yes, for a while I thought it was going to get puppied.

**Comment by shaw05** [comment]
Has anyone mentioned that this would be a fantastic function to merge into the iNav fork of cleanflight? There is an issue which relates to this already: https://github.com/iNavFlight/inav/issues/199

**Comment by gaelj** [comment]
I know. Betaflight needs it too. I still have some new LEDstrip features I would like to add in Cleanflight, for which discussion is still ongoing (such as battery level, see #2160). 
I am willing to work on the merge for inav & etaflight once LEDstrip is fully finished in Cleanflight.
If someone else wants to already merge the existing updates, that's also fine by me :)
Any help on the questions raised in #2160 would be good as well.
